clients/src/main/java/org/apache/kafka/common/utils/ByteBufferOutputStream.java

```diff
@@ -105,7 +105,7 @@ public int initialCapacity() {
         return initialCapacity;
     }
 
-    private void maybeExpandBuffer(int remainingRequired) {
+    public void maybeExpandBuffer(int remainingRequired) {
```

Comment:
I would like to use `DefaultRecordBatch.writeHeader`, which accepts `ByteBuffer`. In order to do that, I need access to the underlying buffer and I need to ensure there is enough room in it. First option is to use `position()` as a hack to expand the buffer:
```java
int initialPosition = bufferOutputStream.position();
bufferOutputStream.position(initialPosition + RECORD_BATCH_OVERHEAD);
bufferOutputStream.position(initialPosition);
writeHeader(bufferOutputStream.buffer());
```
Alternative, what I did in this patch:
```java
bufferOutputStream.maybeExpandBuffer(RECORD_BATCH_OVERHEAD);
writeHeader(bufferOutputStream.buffer());
```
Seems simpler and more direct, no?